### DISCUSSION TITLE
Fallback logic on client mode (IPv6 -> IPv4 (tcp))

### DISCUSSION THREAD

**Discussion by SKatochimoto** [esnet/iperf#1558] [General]
Hi all! 
Sorry for my English :(

I have the following question:
Is it possible to implement a "fallback" logic in client mode (in case we don't explicitly write the argument "IP protocol version -4 && -6" and we use a hostname)?
So that if we have a failed attempt to connect to the server via IPv6 (connection refused), we automatically try via IPv4?

**Comment by davidBar-On** [comment]
Hi, as far as know iperf3 code, this should be possible.  However, I tried with `-c localhost` on the client side, and once `-4` and once `-6` on the server side, and the test succeeds when the server is set with `-4`.  I.e. it seems that the default is using IPv4 address, so the retry should be with IPv6.  Is it different on your machine?

**Reply by bmah888** [reply]
As I understand this, the question is basically about implementing "Happy Eyeballs" support for iperf3, where it would try different address families (presumably IPv6, then IPv4) until one succeeds, or they all fail.

I understand why this is useful for user-facing applications such as Web browsers. But given iperf3's use as a network tester, and the complexity of the current iperf3 code, it's not really a high priority for our use cases. In fact we generally want to have deterministic behavior with respect to what address family is being used. One could probably simulate this by wrapping iperf3 in some sort of script.

**Comment by davidBar-On** [comment]
Submitted a proposed enhancement #1562 - when server hostname is used by the client (and no `-4` or `-6` options) and connection to the server fails, retry connecting with the other address type.
